core/src/test/scala/com/softwaremill/react/kafka2/ProducerTest.scala

```diff
@@ -0,0 +1,242 @@
+        if (msg.value() == "2") Failure(error)
+        else Success(inputMap(msg))
+      })
+    }
+    val (source, sink) = TestSource
+      .probe[Record]
+      .via(Producer(() => client.mock))
+      .map { f => Await.ready(f, 1 second); f.value.get }
+      .toMat(TestSink.probe)(Keep.both)
+      .run()
+
+    sink.request(100)
+    input.map(_._1).foreach(source.sendNext)
+    source.sendComplete()
+
+    val expectedResult = input.map {
+      case x @ (msg, _) if msg.value() == "2" => Failure(error)
+      case x => Success(x)
+    }
+    sink
+      .expectNextN(expectedResult)
+      .expectComplete()
+
+    client.verifyClosed()
+    client.verifySend(atLeastOnce())
+    client.verifyNoMoreInteractions()
+  }
+
+  it should "close client and complete in case of cancellation of outlet" in {
+    val input = 1 to 3 map recordAndMetadata
+
+    val client = {
+      val inputMap = input.toMap
+      new ProducerMock[K, V](ProducerMock.handlers.delayedMap(1 second)(x => Try{ inputMap(x) }))
+    }
+    val (source, sink) = TestSource
+      .probe[Record]
+      .via(Producer(() => client.mock))
+      .mapAsync(1)(identity)
+      .toMat(TestSink.probe)(Keep.both)
+      .run()
+
+    sink.request(10)
+    input.map(_._1).foreach(source.sendNext)
+
+    sink.cancel()
+    source.expectCancellation()
+
+    client.verifyClosed()
+    ()
+  }
+}
+
+object ProducerMock {
+  type Handler[K, V] = (ProducerRecord[K, V], Callback) => Future[RecordMetadata]
+  object handlers {
+    def fail[K, V]: Handler[K, V] = (_, _) => throw new Exception("Should not be called")
+    def delayedMap[K, V](delay: FiniteDuration)(f: ProducerRecord[K, V] => Try[RecordMetadata])(implicit as: ActorSystem): Handler[K, V] = {
+      (record, _) =>
+        implicit val ec = as.dispatcher
+        val promise = Promise[RecordMetadata]()
+        as.scheduler.scheduleOnce(delay) {
+          promise.complete(f(record))
+          ()
+        }
+        promise.future
+    }
+  }
+}
+
+class ProducerMock[K, V](handler: ProducerMock.Handler[K, V])(implicit ec: ExecutionContext) {
+  var closed = false
+  val mock = {
+    val result = Mockito.mock(classOf[KafkaProducer[K, V]])
+    Mockito.when(result.send(any[ProducerRecord[K, V]], any[Callback])).thenAnswer(new Answer[java.util.concurrent.Future[RecordMetadata]] {
+      override def answer(invocation: InvocationOnMock) = {
+        val record = invocation.getArguments()(0).asInstanceOf[ProducerRecord[K, V]]
+        val callback = invocation.getArguments()(1).asInstanceOf[Callback]
+        handler(record, callback).onComplete {
+          case Success(value) if !closed => callback.onCompletion(value, null)
+          case Success(value) if closed => callback.onCompletion(null, new Exception("Kafka producer already closed"))
+          case Failure(ex: Exception) => callback.onCompletion(null, ex)
+          case Failure(ex) => ???
```

Comment:
Is this for the `Throwable` case when Throwable is not an Exception? Maybe this should be called `case Failure(throwableUnsupported)` or something similar to indicate that it's not some unfinished code.